### DISCUSSION TITLE
Making (micro)payments possible via the CLI

### DISCUSSION THREAD

**Discussion by bnuyts** [npm/feedback#134] [CLI] [answered]
This suggestion is about implementing funding or payment streams directly into the npm cli.

**Problem Statement**
Projects deserve their spotlight and request for funding. I think we're all familiar with the npm fund command and I really like it! However, I would like to show my appreciation to all the packages that make my projects possible, and not just a few.

**Potential Solution**
Micropayments make it possible to pay in fraction of cents, and thus also regularly. The motivation behind this is that I'm more willing to share a cent or more each time that I build or release my project with the dependency package authors, then make a donation each quarter or so when I see fit.

Last week I've setup a small PoC for this idea using a cryptocurrency, XRP, to scan the node_modules package.json files for a donate property, collect the donation addresses and split the amount I'm willing to share over those that are willing to receive a donation.

This flow could run on each release or CI build. You can find an example [here](https://github.com/bnuyts/xrp-payment-cli-example)

Now, where does npm come into play? I think it would be great if there was a standard for the donation property in the package.json file. It could define multiple currencies, wallets, etc. npm could also foresee a standard in how a payment from the terminal could be triggered without the need of adding a global or devdependency on another package.

**Comment by MylesBorins** [comment]
Hey @bnuyts

There is some prior art of discussion about this over [in the cli RFC repo](https://github.com/npm/rfcs/pull/197). Microtransactions for modules is definitely a novel and noble idea... but I am personally skeptical of how it can be accomplished. One important challenge with something like this is figuring out how to appropriately allocate funds. While "install" does indeed seem like a good metric, it reflects usage in development (install many times with CI + local) and doesn't reflect usage in production (deployed once and lives for an extended period of time).

We do already support a way for folks to specify where donations go to, as you mentioned above. It seems to me like a 3rd party microtransaction service could build on top of that. Another thing to keep in mind is that additional metadata can always be added to a package.json... while the npm client itself does look for certain fields (e.g. npm fund) that doesn't have to limit metadata that is included in the package.json. It seems to me like the ask here is specifically to make it easier to fund directly using npm itself without requiring a 3rd party package... but considering the number of cryptocurrencies, funding sources, and payment methods it seems like the scope for such a feature is extremely broad.

For now, you very well could build out your own "donation" field in the package.json and start with a 3rd party tool that could understand this field and act on it... at the very least a 3rd party tool will likely be needed to figure out how to do the actual transactions as the npm cli doesn't know anything about making payments.

**Comment by joelwass** [comment, marked as answer]
[Flossbank](https://flossbank.com) is built to support donations to entire dep trees that are used via micro payments

As part of doing this, we are currently implementing the "receiving" piece of the equation, i.e. how does someone specify where they want their money sent. Our current plan is to overload the "funding" package.json keyword to allow interledger payment pointers to be put in as a value, allowing us (and others) to make micro transactions through whatever means they'd like to the payment pointer. We think normalizing the standard that anyone can put in payment destinations to their package.json would be beneficial for everyone, as it'd allow any third parties such as ourselves to build systems around this standard and not depend on some proprietary system, whether it be ours or others.

Figuring out how to "allocate" funds down dep tree's is also incredibly interesting. I'd love to bounce ideas around, we've found many people are interested in finding a solution. Our current implementation can be found [here](https://flossbank.com/how-it-works) but we're assuming a better solution is out there, as this one is naive and intended to protect against package cycle abuse of our system, so i'm sure ideally there are better solutions.

**Reply by bnuyts** [reply]
@joelwass sounds like a great project and is pretty much what I had in mind 👍
